chemprop/uncertainty/evaluator.py

```diff
@@ -41,14 +44,68 @@ def evaluate(self, preds: Tensor, uncs: Tensor, targets: Tensor, mask: Tensor) -
+    calibration on an observed-probability versus expected-probability plot.
+    """
+
     def evaluate(self, preds: Tensor, uncs: Tensor, targets: Tensor, mask: Tensor) -> Tensor:
-        return
+        bin_scaling = torch.tensor([erfinv(i / 100) * np.sqrt(2) for i in range(1, 100)])
+        bin_scaling = bin_scaling.view(-1, 1, 1)
+        errors = torch.abs(preds - targets)
+        uncs = torch.sqrt(uncs).unsqueeze(0)
+        bin_unc = uncs * bin_scaling
+        bin_count = bin_unc >= errors.unsqueeze(0)
+        mask = mask.unsqueeze(0)
+        bin_fraction = (bin_count & mask).sum(1) / mask.sum(1)
+        fractions = torch.zeros((uncs.shape[-1], 101))
+        fractions[:, 100] = 1.0
+        fractions[:, 1:100] = bin_fraction.permute(1, 0)
+        miscal_area = torch.sum(0.01 * torch.abs(fractions - torch.arange(101) / 100), dim=1)
+        return miscal_area
 
 
 @UncertaintyEvaluatorRegistry.register("ence")
 class ExpectedNormalizedErrorEvaluator(RegressionEvaluator):
+    r"""
+    A class that evaluates uncertainty performance by binning together clusters of predictions
+    and comparing the average predicted variance of the clusters against the RMSE of the cluster. [1]_
+
+    .. math::
+        ENCE = \frac{1}{K} \sum_{i=1}^{K} \frac{|\text{mVAR}(i) - RMSE(i)|}{\text{mVAR}(i)}
+
+    where :math:`mVAR(i)` is the root of the mean uncertainty over the :math:`i`-th bin and :math:`RMSE(i)`
+    is the root mean square error over the :math:`i`-th bin.This discrepancy is further normalized by the
+    uncertainty overthe bin, :math:`mVAR(i)`, because the error is expected to be naturally higher as the uncertainty increases.
+
+    References
+    ----------
+    .. [1] Scalia, G.; Grambow, C. A.; Pernici, B.; Li, Y. P.; Green, W. H. "Evaluating Scalable Uncertainty Estimation Methods for Deep Learning-Based Molecular Property Prediction." J. Chem. Inf. Model., 2020, 60(6), 2697-2717. https://pubs.acs.org/doi/10.1021/acs.jcim.9b00975.
+    """
+
     def evaluate(self, preds: Tensor, uncs: Tensor, targets: Tensor, mask: Tensor) -> Tensor:
-        return
+        root_mean_vars = torch.zeros([uncs.shape[1], 100])
+        rmses = torch.zeros_like(root_mean_vars)
+
+        for j in range(uncs.shape[1]):
+            mask_j = mask[:, j]
+            preds_j = preds[:, j][mask_j]
+            targets_j = targets[:, j][mask_j]
+            uncs_j = uncs[:, j][mask_j]
+            errors = torch.abs(preds_j - targets_j)
+
+            sort_idx = torch.argsort(uncs_j)
+            uncs_j = uncs_j[sort_idx]
+            errors = errors[sort_idx]
+
+            split_unc = torch.chunk(uncs_j, 100)
+            split_error = torch.chunk(errors, 100)
+
+            root_mean_vars[j] = torch.tensor([torch.sqrt(torch.mean(chunk)) for chunk in split_unc])
+            rmses[j] = torch.tensor([torch.sqrt(torch.mean(chunk**2)) for chunk in split_error])
+
+        ence = torch.mean(torch.abs(root_mean_vars - rmses) / root_mean_vars, axis=1)
```

Comment:
is there a way to write this using broadcasting? I can give it a shot later, but you know what's going on here better than I do